### PR TITLE
Create a standard failure object for the API to hand back to the UI project

### PR DESCRIPTION
**Pull request by gitCarrot**
# Ticket Link

[Ticket 1538](https://uhawaii.atlassian.net/browse/GROUPINGS-1538)

# List of squashed commits

- Create ExceptionForTesting exception with ApiError and SubError for endpoint /exception
- Add maven library for timestamp JsonFormat (Jackson)
- Create Sub Error to contain the error data of service layer to be used in UI side
- Create ApiValidationError for the data to be set in the subError objecdt
- Create Test for RestExceptionHandler
- Create RestExceptionHandler for handling global error from dispatcher
- Change the throwable object with ApiError and SubApiError at the /exception endpoint
- Make the test for expected json from the endpoint /exception
- Modify the content of the error message for ExceptionForTesting exception
- Create the test for the type ApiError (Standard Failure Object)
- Add instance of List<ApiSubError> subErrors for contain the sub error on the error response
- Clean up the function of future ticket for refactoring the exception handling, just leave the exception for test and one of the Exception that is thrown from service layer
- Add the tests for checking the json content of the ApiError object from RestExceptionHandler at the controller and service layer.
- Add the function 'addAllSubErrors' for clean code
- Merge the functions from RestExceptionHandler to ErrorControllerAdvice
- Modify existed error handling functions in ErrorControllerAdvice
- Remove RestExeptionHandlerTest
- Fix spelling error
- Put message variable and its getters and setters into the ApiSubError abstract class
- Move all the tests from RestExceptionHanlderTest to ErrorControllerAdviceTest and remove the file of RestExceptionHanlderTest
- Modify the result of the status in regex test from ClIENT_ERROR to SERVER_ERROR
- Add comment for custom exception test section
- Remove JsonFormat library for timestamp at ApiError
- Remove maven library jackson-datatype-jsr310 for timestamp
- Add the logic for verifying whether the subErrors list passed to addAllSubErrors is empty
- Set the return type to ReponseEntity<ApiError>
- Set PropertyLocator as local variable
- Remove import for codacy
- Include assert() for JUnit test
- Remove PropertyLocator before making new integration test
- Writing a unit test for ApiValidationError
- Remove comments and * from import
- Remove * import and use single class import

# Test Checklist

- [x] Exhibits Clear Code Structure:
- [x] Project Unit Tests Passed:
- [x] Project Integration Tests Passed:
- [x] Executes Expected Functionality:
- [x] Tested For Incorrect/Unexpected Inputs:

### CODE REVIEW

**Comment by JorWo** [comment]
There are some Codacy issues to fix. Also you mentioned that Jasmine tests are not working and that I am assuming you are trying to run them in the API project. The API project does not have Jasmine tests, only the UI. So you are good.

File: pom.xml [hunk 1]
```diff
@@ -50,6 +50,10 @@
             <artifactId>opencsv</artifactId>
             <version>5.7.1</version>
         </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.datatype</groupId>
+            <artifactId>jackson-datatype-jsr310</artifactId>
+        </dependency>
```
Comment (JorWo) [inline pom.xml:56]: Could you check if this dependency is needed? I remember using this dependency for datetime but finding out it is not needed.

File: src/main/java/edu/hawaii/its/api/controller/GroupingsRestControllerv2_1ForTesting.java
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/controller/GroupingsRestControllerv2_1ForTesting.java:1]: I have been looking at this file's long name and I think it deservers a rename. What do you think about AssessmentRestController or ForTestingRestController?

File: src/test/java/edu/hawaii/its/api/controller/RestExceptionHandlerTest.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
+package edu.hawaii.its.api.controller;
+
+import edu.hawaii.its.api.configuration.SpringBootWebApplication;
+import edu.hawaii.its.api.exception.AccessDeniedException;
+import edu.hawaii.its.api.exception.UhMemberNotFoundException;
+import edu.hawaii.its.api.service.*;
```
Comment (JorWo) [inline src/test/java/edu/hawaii/its/api/controller/RestExceptionHandlerTest.java:6]: Make sure to remove any * imports, thanks.

File: src/main/java/edu/hawaii/its/api/type/ApiSubError.java
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/type/ApiSubError.java:1]: I see your goal with the ApiSubError abstract class, to create a bunch of sub error classes to extend it. Abstract classes are great for requiring all classes that extend it to contain already implemented functions and variables. However, this abstract has nothing in it. I think the `message` variable and its getters and setters should be put into this abstract class.

File: src/main/java/edu/hawaii/its/api/controller/RestExceptionHandler.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                .timestamp(LocalDateTime.now());
+
+        List<ApiSubError> subErrors = ex.getSubErrors();
+
+        /* Add all SubErrors in the ApiError */
+        if(subErrors != null) {
+            for (ApiSubError subError : subErrors) {
+                errorBuilder.addSubError(subError);
+            }
+        }
```
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/controller/RestExceptionHandler.java:90]: This null guard if statement is not needed, the for loop will not run in the subErrors List is empty.

File: src/main/java/edu/hawaii/its/api/controller/RestExceptionHandler.java
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/controller/RestExceptionHandler.java:1]: Could you remove the comments: `// Other custom exception handlers using Builder pattern`, `/* Add all SubErrors in the ApiError */`, and `/* BUILD */`? It's good to write comments but it is not always needed when the code can explain itself. I also liked that you added comments to explain the `handleExceptionForTesting` function. For the rest of the multi-line comments, be sure to format them as one comment like this:
```
/*
    Common Error Message from @ExceptionForTesting.class
    status, message, timestamp must be nonNull
*/
```
Thanks!

**Comment by JorWo** [comment]
Make sure  that you have written a unit/integration test for all the classes you created in this PR and that its test coverage is 100% or as close to 100% as possible. Check the test coverage using JaCoCo. More information in our bwiki: https://uhawaii.atlassian.net/wiki/spaces/SITARd/pages/12943618/jaCoCo+Code+Coverage

File: src/main/java/edu/hawaii/its/api/controller/RestExceptionHandler.java (hunk 3 above)
Comment (gitCarrot) [reply src/main/java/edu/hawaii/its/api/controller/RestExceptionHandler.java:90]: If the 86th line is absent, a 'NullPointerException' was occurring in an existing exception used within the service without the previous 'subError', so I made it.

File: src/test/java/edu/hawaii/its/api/controller/RestExceptionHandlerTest.java (hunk 2 above)
Comment (gitCarrot) [reply src/test/java/edu/hawaii/its/api/controller/RestExceptionHandlerTest.java:6]: Thanks for letting me know

File: src/main/java/edu/hawaii/its/api/controller/GroupingsRestControllerv2_1ForTesting.java
Comment (gitCarrot) [reply src/main/java/edu/hawaii/its/api/controller/GroupingsRestControllerv2_1ForTesting.java:1]: ForTestingRestController sounds good! thanks

File: src/main/java/edu/hawaii/its/api/controller/RestExceptionHandler.java
Comment (gitCarrot) [reply src/main/java/edu/hawaii/its/api/controller/RestExceptionHandler.java:1]: Alright! I got it

File: src/main/java/edu/hawaii/its/api/type/ApiSubError.java
Comment (gitCarrot) [reply src/main/java/edu/hawaii/its/api/type/ApiSubError.java:1]: You're right. I fixed it

File: src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java [hunk 4]
```diff
@@ -27,63 +31,165 @@ public class ErrorControllerAdvice {
     @Autowired
     private EmailService emailService;
 
+    private ResponseEntity<Object> buildResponseEntity(ApiError apiError) {
+        return new ResponseEntity<>(apiError, apiError.getStatus());
+    }
+
     @ExceptionHandler(GroupingsServiceResultException.class)
-    public ResponseEntity<GroupingsHTTPException> handleGroupingsServiceResultException(
+    public ResponseEntity<Object> handleGroupingsServiceResultException(
```
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java:39]: If possible, could you set the return type to "ReponseEntity<ApiError>" for all of the functions in here? Thanks.

File: pom.xml (hunk 1 above)
Comment (gitCarrot) [reply pom.xml:56]: I think we don't really need it. I just removed it!

File: src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java (hunk 4 above)
Comment (JorWo) [reply src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java:39]: Sorry, I formatted my message wrong and made `<ApiError>` invisible.

File: src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java [hunk 5]
```diff
@@ -27,63 +31,165 @@ public class ErrorControllerAdvice {
...
+
+        ApiError.Builder errorBuilder = new ApiError.Builder()
+                .status(HttpStatus.INTERNAL_SERVER_ERROR)
+                .message("Validation failed")
+                .debugMessage("Check your service at endpoint /exception")
+                .timestamp(LocalDateTime.now());
+
+        if (ex.getSubErrors() != null) {
+            errorBuilder.addAllSubErrors(ex.getSubErrors());
+        }
```
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java:171]: Could you put this null guard inside the errorBuilder.addAllSubErrors function? That we do not need to repeat this null guard for every exception we create. Thanks!

**Comment by JorWo** [comment]
Could you also write a unit test for ApiValidationError? Thanks.

File: src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java [hunk 6]
```diff
@@ -1,24 +1,28 @@
 package edu.hawaii.its.api.controller;
 
 import java.io.IOException;
+import java.time.LocalDateTime;
 
+import edu.hawaii.its.api.exception.*;
```
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java:6]: Make sure to remove the * import here.

File: src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java [hunk 7]
```diff
@@ -27,63 +31,161 @@ public class ErrorControllerAdvice {
...
+                .timestamp(LocalDateTime.now());
+
+        ApiError apiError = errorBuilder.build();
+
+        return buildResponseEntity(apiError);
     }
 
-    private ResponseEntity<GroupingsHTTPException> exceptionResponse(String message, Throwable cause, int status) {
-        GroupingsHTTPException httpException = new GroupingsHTTPException(message, cause, status);
+    /* Custom Exception for UH GROUPINGS */
```
Comment (JorWo) [inline src/main/java/edu/hawaii/its/api/controller/ErrorControllerAdvice.java:157]: This comment can be removed too.

**Review by mhodgesatuh — COMMENTED** [review]
public class ForTestingRestControllerTest is redundantly named.  The "Test" suffix indicates that the class is for testing. RestControllerTest would be the succinct name.  However, would it be more accurate to name it the ErrorControllerTest class?

Frank should be a reviewer for this particular ticket.

**Comment by mhodgesatuh** [comment]
Perhaps I'm wrong about the class name, but the discussion appears to be worthwhile.

**Comment by JorWo** [comment]
As of now, it does not seem that we will add anything more to the ForTestingController, so I'm suggesting that we should put everything back into the GroupingsRestController. Eddy, you can keep the rename change in this PR and I'll create a ticket for someone to handle moving the endpoint from ForTestingController to GroupingsRestController.

**Comment by JorWo** [comment]
Sorry, I just found more * imports in the following files: ErrorControllerAdviceTest, ApiErrorTest, ApiValidationErrorTest. Could you remove them? Thanks!

**Pull request merged**